contrib/src/databases.rs

```diff
@@ -0,0 +1,84 @@
+pub extern crate r2d2;
+
+use std::collections::BTreeMap;
+use std::error::Error;
+use std::io;
+use std::marker::{Send, Sized};
+use rocket::config::{self, ConfigError, Value};
+
+use self::r2d2::ManageConnection;
+
+#[cfg(feature = "diesel_sqlite")]
+pub extern crate diesel;
+
+#[cfg(feature = "pg_pool")]
+pub extern crate postgres;
+#[cfg(feature = "pg_pool")]
+pub extern crate r2d2_postgres;
+
+pub struct DatabaseConfig<'a> {
+    pub url: &'a str,
+    pub pool_size: u32,
+    pub extras: Option<&'a BTreeMap<String, Value>>,
+}
+
+pub fn database_config<'a>(name: &str, from: &'a config::Config) -> Option<DatabaseConfig<'a>> {
```

Comment:
So, I get that this entire method can be hard to follow since I'm using the monadic interface of Options and Results instead of unwrapping/lifting the type. I think this is the best solution, though, since a lot of these returned types don't allow for easy use of the `?`.

Additionally, I simplified the pattern matching on line 33. I also converted the `DatabaseConfig` `extras` field to an `Option<&'a BTreeMap<String, Value>>` to allow for it to not exist and not require an `unwrap` or pattern match.

Reply:
Argh, yeah, I'm still not super happy with this function. Let's not worry about it for now; the interface looks okay, and we can work on improving the implementation later.

Reply:
Yeah, it's a little messy. We can revisit it before we end up merging this work. There are just a lot of fail-able parts that need to propagate the error upwards.